---
title: "2DE_Heatmap"
author: "Jaakko"
date: "2025-11-20"
format:
  html:
    css: /styles.css
    code-copy: true 
    code-line-numbers: true
    code-overflow: scroll
    code-tools: true
    highlight-style: github-dark
engine: knitr
execute:
  echo: true
  eval: true
conda:
  environment: bioinfo #bioinfo_environment.yml
---

## Quarto

This is a Quarto document. Quarto is a multi-language, next-generation version of R Markdown. Check the [tutorial](https://quarto.org/docs/get-started/hello/rstudio.html).\
N.B. - Learn the IT foundations to master bioinformatics tools and softwares, and prevent getting tangled on the infamous "dependency hell". Ask for the "Foundations of Bioinformatics Infrastructure".

## Create and activate the environment "bioinfo"

```{bash}
#| eval: false
mamba create -n bioinfo python=3.11
mamba activate bioinfo
```

## Download the RNA-seq dataset

Let's use a real example - [*RNA-seq of ITPR3 and RELB knockout in SW480 under normoxia and hypoxia*]{.underline} (*Homo sapiens*) <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576>.

![](images/IM1.png)

N.B. - Structure properly your folder ('mkdir folder'):\

```{bash}
#| echo: false
cd ..
printf ".\n├── data/\n├── scripts/\n└── results/\n"
```

How to download the files from FTP with UNIX terminal: <https://www.ncbi.nlm.nih.gov/geo/info/download.html>\
<https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/>

```{bash}
#| eval: false
# Move to /data folder !!!
cd data
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/GSE197576_raw_gene_counts_matrix.tsv.gz
```

Alternative use GEOquery <https://bioconductor.org/packages/release/bioc/html/GEOquery.html>.

## Check the data ad extract the samples of interest

Check the data in the compressed RNA-seq file.

```{bash}
#| eval: false
zless -S GSE197576_raw_gene_counts_matrix.tsv.gz
# z because is compressed
# less is the faster command to view the content of a file
# -S maintain one raw column
# press q to quit
```

Install csvtk annd check the data in a pretty aligned manner.

```{bash}
#| eval: false
mamba install -c bioconda csvtk
```

```{bash}
#| eval: false
csvtk pretty -t ../data/GSE197576_raw_gene_counts_matrix.tsv.gz | less -S
# -t tab the limit
# press q to quit
```

```{bash}
#| echo: true
csvtk headers -t ../data/GSE197576_raw_gene_counts_matrix.tsv.gz 
```

N.B - Get csvtk at <https://github.com/shenwei356/csvtk>

Extract the gene expression values from:

|  |  |  |  |
|----|----|----|----|
| *01_SW_sgCTRL_Norm* | *02_SW_sgCTRL_Norm* | *17_SW_sgRELB_3_Hyp* | *18_SW_sgRELB_3_Hyp* |

```{bash}
csvtk cut -t -f1,2,3,12,13 ../data/GSE197576_raw_gene_counts_matrix.tsv.gz | column -t | head
```

```{bash}
#| eval: false
csvtk cut -t -f1,1,2,12,13 ../data/GSE197576_raw_gene_counts_matrix.tsv.gz > raw.counts.tsv
```

Your */data* folder now contains:

```{bash}
#| echo: false
cd ..
tree -n -F -L 1 --noreport data 
```

## Read the data into R and make a DESeq2 object

Follow the tutorial <http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>.

### Setup

Install the R-packages for this project within the 'bioinfo' virtual environment.

```{bash}
#| eval: false
mamba install -c conda-forge r-tidyverse
mamba install -c conda-forge r-dplyr
mamba install -c conda-forge r-here
mamba install -c bioconda bioconductor-deseq2
mamba intall -c conda-forge r-ggplot2
mamba install -c conda-forge r-ggrepel
mamba install -c conda-forge r-BiocManager
mamba install -c bioconda bioconductor-complexheatmap
```

Save the environment for reproducibility.

```{bash}
#| eval: false
mamba env export --no-builds > bioinfo_environment.yml
```

Open RStudio from the terminal: you are now working with the R software within 'bioinfo' and you can find the packages previously installed.

```{bash}
#| eval: false
open -na RStudio
```

### Let’s start

```{r}
#| message: false
#| warning: false
library(dplyr) #https://dplyr.tidyverse.org/
library(readr) #https://readr.tidyverse.org/
library(here) #https://cran.r-project.org/web/packages/here/vignettes/here.html
library(DESeq2) #https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8
library(ggplot2)
library(ggrepel)
library(ComplexHeatmap)

raw_counts <- read_tsv(here("01_Heatmap_Genomics/2_DE_Heatmap/data/raw.counts.tsv")) #import the dataset

raw_counts_mat <- raw_counts[, -1] %>% as.matrix #create the values matrix removing the first column for DESeq2
head(raw_counts_mat)

rownames(raw_counts_mat) <- raw_counts$gene #add row names to the matrix
head(raw_counts_mat)
```

Make a dataframe for the metadata.

```{r}
#| message: false
#| warning: false
metaframe <- tibble(condition = c ("normoxia", "normoxia", "hypoxia", "hypoxia"))
rownames(metaframe) <- colnames(raw_counts_mat)
metaframe
```

Make a DESeq2 object ([tutorial](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)).

```{r}
#| message: false
#| warning: false
all(rownames(metaframe) == colnames(raw_counts_mat)) #check

dds <- DESeqDataSetFromMatrix(countData = raw_counts_mat, #input object with design
                              colData = metaframe,
                              design = ~ condition)
dds <- DESeq(dds) #DE analysis

?results
res <- results(dds, contrast = c("condition", "hypoxia", "normoxia")) #extract results from the analysis 

res %>% 
  as.data.frame() %>%
  arrange(padj <= 0.1, abs(log2FoldChange) >= 2) %>% #filter significant and substantial expression changes 
  head(n = 10)

significant_genes <- res %>% #identify the respective genes
  as.data.frame() %>%
  filter(padj <= 0.1, abs(log2FoldChange) >= 2) %>%
  rownames()

significant_genes[1:10]
```

### PCA Analysis to verify your results

```{r}
#| message: true
#| warning: false
vsd <- vst(dds, blind = F) #normalization applying a variance stabilizing transformation (VST)

p <- plotPCA(vsd, intgroup = c("condition"))
p + coord_cartesian(ylim = c(-0.7, 0.7))
```

You can make a PCA by yourself.

```{r}
#| warning: false
# vsd <- vst(dds, blind = F)
head(assay(vsd), 3)
?assay

normalized_vsd <- assay(vsd) %>% #from DeSeq object to matrix
 as.matrix()

pca_prcomp <- prcomp(t(normalized_vsd), center = T, scale. = F) #PCA analysis

names(pca_prcomp)
pca_prcomp$x

PC1_PC2 <- data.frame(
  PC1 = pca_prcomp$x[,1],
  PC2 = pca_prcomp$x[,2],
  type = rownames(pca_prcomp$x)
)

var_explained <- (pca_prcomp$sdev^2) / sum(pca_prcomp$sdev^2)
pc1_percent <- round(var_explained[1] * 100, 1)
pc2_percent <- round(var_explained[2] * 100, 1)

ggplot(PC1_PC2, aes(x = PC1, y = PC2, col = type)) +
  geom_point(size = 4) +
  geom_text_repel(
    aes(label = type),
    box.padding = 0.5,
    point.padding = 0.3,
    segment.size = 0.3,
    min.segment.length = 0
  ) +
  
  labs(
    x = paste0("PC1 (", pc1_percent, "% variance)"),
    y = paste0("PC2 (", pc2_percent, "% variance)")
  ) +
  
  coord_cartesian(clip = "off") + 
  theme_bw(base_size = 14) +
  theme(
    plot.margin = margin(10, 5, 10, 5),  
    legend.position = "right"
  )
```

It is not exactly the same, what's going on?

```{r}
#| eval: false
?plotPCA #using the top 500 most variable genes
```

## Make a perfect heatmap

```{r}
significant_mat <- normalized_vsd[significant_genes, ]
dim(significant_mat)

Heatmap(t(scale(t(significant_mat))))

```

You get this perfect looking heatmap because you select the genes that are different. So, no surprise at all!

```{r}
metaframe

meta_annot <- HeatmapAnnotation(df = metaframe,
                                col = list(condition = c("hypoxia" = "red", "normoxia" = "blue")))

Heatmap(t(scale(t(significant_mat))),
        top_annotation = meta_annot,
        show_row_names = F,
        name = "scaled normalized\nexpression"
        )
```

why scaling is important?

```{r}
#| message: false
#| warning: false
Heatmap(significant_mat, 
        top_annotation = meta_annot,
        show_row_names = FALSE,
        name = "NOT scaled normalized\nexpression"
        )
```
