<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jaakko">
<meta name="dcterms.date" content="2025-11-20">

<title>2DE_Heatmap ‚Äì - - - - - - LearningByDoing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-7ac2f9da8c2617a4fdd15004b4601015.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-6346acc0ce68ab0a0a65d2b072e3ba70.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../../styles.css">
</head>

<body class="nav-sidebar floating nav-fixed fullcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">- - - - - - LearningByDoing</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../../01_Heatmap_Genomics/1_ComplexHeatmap/1ComplexHeatmap.html">Heatmap</a></li><li class="breadcrumb-item"><a href="../../../01_Heatmap_Genomics/2_DE_Heatmap/scripts/2DE_Heatmap.html">DE Heatmap</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Heatmap</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../01_Heatmap_Genomics/1_ComplexHeatmap/1ComplexHeatmap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ComplexHeatmap</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../01_Heatmap_Genomics/2_DE_Heatmap/scripts/2DE_Heatmap.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">DE Heatmap</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../../01_Heatmap_Genomics/1_ComplexHeatmap/1ComplexHeatmap.html">Heatmap</a></li><li class="breadcrumb-item"><a href="../../../01_Heatmap_Genomics/2_DE_Heatmap/scripts/2DE_Heatmap.html">DE Heatmap</a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">2DE_Heatmap</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jaakko </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 20, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="quarto" class="level2">
<h2 class="anchored" data-anchor-id="quarto">Quarto</h2>
<p>This is a Quarto document. Quarto is a multi-language, next-generation version of R Markdown. Check the <a href="https://quarto.org/docs/get-started/hello/rstudio.html">tutorial</a>.<br>
N.B. - Learn the IT foundations to master bioinformatics tools and softwares, and prevent getting tangled on the infamous ‚Äúdependency hell‚Äù. Ask for the ‚ÄúFoundations of Bioinformatics Infrastructure‚Äù.</p>
</section>
<section id="create-and-activate-the-environment-bioinfo" class="level2">
<h2 class="anchored" data-anchor-id="create-and-activate-the-environment-bioinfo">Create and activate the environment ‚Äúbioinfo‚Äù</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> bioinfo python=3.11</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="ex">mamba</span> activate bioinfo</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="download-the-rna-seq-dataset" class="level2">
<h2 class="anchored" data-anchor-id="download-the-rna-seq-dataset">Download the RNA-seq dataset</h2>
<p>Let‚Äôs use a real example - <u><em>RNA-seq of ITPR3 and RELB knockout in SW480 under normoxia and hypoxia</em></u> (<em>Homo sapiens</em>) <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576</a>.</p>
<p><img src="images/IM1.png" class="img-fluid"></p>
<p>N.B. - Structure properly your folder (‚Äòmkdir folder‚Äô):<br>
</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>.
‚îú‚îÄ‚îÄ data/
‚îú‚îÄ‚îÄ scripts/
‚îî‚îÄ‚îÄ results/</code></pre>
</div>
</div>
<p>How to download the files from FTP with UNIX terminal: <a href="https://www.ncbi.nlm.nih.gov/geo/info/download.html" class="uri">https://www.ncbi.nlm.nih.gov/geo/info/download.html</a><br>
<a href="https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/" class="uri">https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/</a></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># Move to /data folder !!!</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="bu">cd</span> data</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="fu">wget</span> https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/GSE197576_raw_gene_counts_matrix.tsv.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Alternative use GEOquery <a href="https://bioconductor.org/packages/release/bioc/html/GEOquery.html" class="uri">https://bioconductor.org/packages/release/bioc/html/GEOquery.html</a>.</p>
</section>
<section id="check-the-data-ad-extract-the-samples-of-interest" class="level2">
<h2 class="anchored" data-anchor-id="check-the-data-ad-extract-the-samples-of-interest">Check the data ad extract the samples of interest</h2>
<p>Check the data in the compressed RNA-seq file.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">zless</span> <span class="at">-S</span> GSE197576_raw_gene_counts_matrix.tsv.gz</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co"># z because is compressed</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># less is the faster command to view the conte of a file</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># -S maintain one raw column</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co"># press q to quit</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Install csvtk annd check the data in a pretty aligned manner.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> bioconda csvtk</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">csvtk</span> pretty <span class="at">-t</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co"># -t tab the limit</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co"># press q to quit</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">csvtk</span> headers <span class="at">-t</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>gene
01_SW_sgCTRL_Norm
02_SW_sgCTRL_Norm
03_SW_sgITPR3_1_Norm
04_SW_sgITPR3_1_Norm
07_SW_sgRELB_3_Norm
08_SW_sgRELB_3_Norm
11_SW_sgCTRL_Hyp
12_SW_sgCTRL_Hyp
13_SW_sgITPR3_1_Hyp
14_SW_sgITPR3_1_Hyp
17_SW_sgRELB_3_Hyp
18_SW_sgRELB_3_Hyp</code></pre>
</div>
</div>
<p>N.B - Get csvtk at <a href="https://github.com/shenwei356/csvtk" class="uri">https://github.com/shenwei356/csvtk</a></p>
<p>Extract the gene expression values from:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
</colgroup>
<tbody>
<tr class="odd">
<td><em>01_SW_sgCTRL_Norm</em></td>
<td><em>02_SW_sgCTRL_Norm</em></td>
<td><em>17_SW_sgRELB_3_Hyp</em></td>
<td><em>18_SW_sgRELB_3_Hyp</em></td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-f1,2,3,12,13</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>gene                       01_SW_sgCTRL_Norm  02_SW_sgCTRL_Norm  17_SW_sgRELB_3_Hyp  18_SW_sgRELB_3_Hyp
DDX11L1                    0                  0                  0                   0
WASH7P                     18                 11                 30                  25
MIR6859-1                  5                  1                  8                   5
MIR1302-2HG                0                  0                  0                   0
MIR1302-2                  0                  0                  0                   0
FAM138A                    0                  0                  0                   0
OR4F5                      0                  0                  0                   0
LOC100996442               9                  3                  15                  20
LOC729737                  3                  3                  33                  45</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-f1,1,2,12,13</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz <span class="op">&gt;</span> raw.counts.tsv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Your <em>/data</em> folder now contains:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>data/
‚îú‚îÄ‚îÄ GSE197576_raw_gene_counts_matrix.tsv.gz
‚îî‚îÄ‚îÄ raw.counts.tsv</code></pre>
</div>
</div>
</section>
<section id="read-the-data-into-r-and-make-a-deseq2-object" class="level2">
<h2 class="anchored" data-anchor-id="read-the-data-into-r-and-make-a-deseq2-object">Read the data into R and make a DESeq2 object</h2>
<p>Follow the tutorial <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a>.</p>
<section id="setup" class="level3">
<h3 class="anchored" data-anchor-id="setup">Setup</h3>
<p>Install the R-packages for this project within the ‚Äòbioinfo‚Äô virtual environment.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-tidyverse</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-dplyr</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-here</span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="ex">mamba</span> install <span class="at">-c</span> bioconda bioconductor-deseq2</span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="ex">mamba</span> intall <span class="at">-c</span> conda-forge r-ggplot2</span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-ggrepel</span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-BiocManager   </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Save the environment for reproducibility.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1"></a><span class="ex">mamba</span> env export <span class="at">--no-builds</span> <span class="op">&gt;</span> bioinfo_environment.yml</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Open RStudio from the terminal: you are now working with the R software within ‚Äòbioinfo‚Äô and you can find the packages previously installed.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="ex">open</span> <span class="at">-na</span> RStudio</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="lets-start" class="level3">
<h3 class="anchored" data-anchor-id="lets-start">Let‚Äôs start</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="fu">library</span>(dplyr) <span class="co">#https://dplyr.tidyverse.org/</span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="fu">library</span>(readr) <span class="co">#https://readr.tidyverse.org/</span></span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="fu">library</span>(here) <span class="co">#https://cran.r-project.org/web/packages/here/vignettes/here.html</span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="fu">library</span>(DESeq2) <span class="co">#https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8</span></span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb16-7"><a href="#cb16-7"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"ComplexHeatmap"</span>)</span>
<span id="cb16-8"><a href="#cb16-8"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb16-9"><a href="#cb16-9"></a></span>
<span id="cb16-10"><a href="#cb16-10"></a>raw_counts <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="fu">here</span>(<span class="st">"01_Heatmap_Genomics/2_DE_Heatmap/data/raw.counts.tsv"</span>)) <span class="co">#import the dataset</span></span>
<span id="cb16-11"><a href="#cb16-11"></a></span>
<span id="cb16-12"><a href="#cb16-12"></a>raw_counts_mat <span class="ot">&lt;-</span> raw_counts[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">%&gt;%</span> as.matrix <span class="co">#create the values matrix removing the first column for DESeq2</span></span>
<span id="cb16-13"><a href="#cb16-13"></a><span class="fu">head</span>(raw_counts_mat)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     01_SW_sgCTRL_Norm 02_SW_sgCTRL_Norm 17_SW_sgRELB_3_Hyp 18_SW_sgRELB_3_Hyp
[1,]                 0                 0                  0                  0
[2,]                18                11                 30                 25
[3,]                 5                 1                  8                  5
[4,]                 0                 0                  0                  0
[5,]                 0                 0                  0                  0
[6,]                 0                 0                  0                  0</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">rownames</span>(raw_counts_mat) <span class="ot">&lt;-</span> raw_counts<span class="sc">$</span>gene <span class="co">#add row names to the matrix</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="fu">head</span>(raw_counts_mat)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>            01_SW_sgCTRL_Norm 02_SW_sgCTRL_Norm 17_SW_sgRELB_3_Hyp
DDX11L1                     0                 0                  0
WASH7P                     18                11                 30
MIR6859-1                   5                 1                  8
MIR1302-2HG                 0                 0                  0
MIR1302-2                   0                 0                  0
FAM138A                     0                 0                  0
            18_SW_sgRELB_3_Hyp
DDX11L1                      0
WASH7P                      25
MIR6859-1                    5
MIR1302-2HG                  0
MIR1302-2                    0
FAM138A                      0</code></pre>
</div>
</div>
<p>Make a dataframe for the metadata.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>metaframe <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">condition =</span> <span class="fu">c</span> (<span class="st">"normoxia"</span>, <span class="st">"normoxia"</span>, <span class="st">"hypoxia"</span>, <span class="st">"hypoxia"</span>))</span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="fu">rownames</span>(metaframe) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(raw_counts_mat)</span>
<span id="cb20-3"><a href="#cb20-3"></a>metaframe</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 √ó 1
  condition
* &lt;chr&gt;    
1 normoxia 
2 normoxia 
3 hypoxia  
4 hypoxia  </code></pre>
</div>
</div>
<p>Make a DESeq2 object (<a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">tutorial</a>).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(metaframe) <span class="sc">==</span> <span class="fu">colnames</span>(raw_counts_mat)) <span class="co">#check</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> raw_counts_mat, <span class="co">#input object with design</span></span>
<span id="cb24-2"><a href="#cb24-2"></a>                              <span class="at">colData =</span> metaframe,</span>
<span id="cb24-3"><a href="#cb24-3"></a>                              <span class="at">design =</span> <span class="sc">~</span> condition)</span>
<span id="cb24-4"><a href="#cb24-4"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds) <span class="co">#DE analysis</span></span>
<span id="cb24-5"><a href="#cb24-5"></a></span>
<span id="cb24-6"><a href="#cb24-6"></a>?results</span>
<span id="cb24-7"><a href="#cb24-7"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"condition"</span>, <span class="st">"hypoxia"</span>, <span class="st">"normoxia"</span>)) <span class="co">#extract results from the analysis </span></span>
<span id="cb24-8"><a href="#cb24-8"></a></span>
<span id="cb24-9"><a href="#cb24-9"></a>res <span class="sc">%&gt;%</span> </span>
<span id="cb24-10"><a href="#cb24-10"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb24-11"><a href="#cb24-11"></a>  <span class="fu">arrange</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.1</span>, <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;=</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="co">#filter significant and substantial expression changes </span></span>
<span id="cb24-12"><a href="#cb24-12"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>               baseMean log2FoldChange     lfcSE       stat    pvalue      padj
WASH7P        19.868995     0.41727500 0.5998607  0.6956198 0.4866669 0.6105208
MIR6859-1      4.356427     0.65513668 1.3035290  0.5025870 0.6152547 0.7232638
LOC100996442  10.646734     1.06292492 0.8520192  1.2475363 0.2122009 0.3229198
WASH9P        50.042666     0.04451088 0.3788408  0.1174923 0.9064699 0.9408732
LINC01409     12.417663     0.80294905 0.7708893  1.0415880 0.2976028 0.4204900
FAM87B         2.852901    -0.33928196 1.5932534 -0.2129492 0.8313666 0.8900030
LINC00115     24.063112    -0.30645340 0.5400642 -0.5674389 0.5704160 0.6854688
LOC100288175 162.124769     0.20237040 0.2175051  0.9304167 0.3521554 0.4798503
LOC105378948  39.581280     0.61997436 0.4277432  1.4494080 0.1472237 0.2390383
RNF223         9.592991    -0.79713059 0.8827150 -0.9030442 0.3665025 0.4942234</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>significant_genes <span class="ot">&lt;-</span> res <span class="sc">%&gt;%</span> <span class="co">#identify the respective genes</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.1</span>, <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;=</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4"></a>  <span class="fu">rownames</span>()</span>
<span id="cb26-5"><a href="#cb26-5"></a></span>
<span id="cb26-6"><a href="#cb26-6"></a>significant_genes[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "LOC729737"    "LINC02593"    "LOC107985728" "LOC107985376" "LINC02781"   
 [6] "LINC01714"    "C1orf167"     "NPPB"         "PADI2"        "PLA2G5"      </code></pre>
</div>
</div>
</section>
<section id="pca-analysis-to-verify-your-results" class="level3">
<h3 class="anchored" data-anchor-id="pca-analysis-to-verify-your-results">PCA Analysis to verify your results</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>vsd <span class="ot">&lt;-</span> <span class="fu">vst</span>(dds, <span class="at">blind =</span> F) <span class="co">#normalization applying a variance stabilizing transformation (VST)</span></span>
<span id="cb28-2"><a href="#cb28-2"></a></span>
<span id="cb28-3"><a href="#cb28-3"></a>p <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(vsd, <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"condition"</span>))</span>
<span id="cb28-4"><a href="#cb28-4"></a>p <span class="sc">+</span> <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.7</span>, <span class="fl">0.7</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2DE_Heatmap_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can make a PCA by yourself.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co"># vsd &lt;- vst(dds, blind = F)</span></span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="fu">head</span>(<span class="fu">assay</span>(vsd), <span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>          01_SW_sgCTRL_Norm 02_SW_sgCTRL_Norm 17_SW_sgRELB_3_Hyp
DDX11L1            10.31070          10.31070           10.31070
WASH7P             10.48764          10.46633           10.51233
MIR6859-1          10.40400          10.35764           10.41488
          18_SW_sgRELB_3_Hyp
DDX11L1             10.31070
WASH7P              10.49477
MIR6859-1           10.39306</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>?assay</span>
<span id="cb31-2"><a href="#cb31-2"></a></span>
<span id="cb31-3"><a href="#cb31-3"></a>normalized_vsd <span class="ot">&lt;-</span> <span class="fu">assay</span>(vsd) <span class="sc">%&gt;%</span> <span class="co">#from DeSeq object to matrix</span></span>
<span id="cb31-4"><a href="#cb31-4"></a> <span class="fu">as.matrix</span>()</span>
<span id="cb31-5"><a href="#cb31-5"></a></span>
<span id="cb31-6"><a href="#cb31-6"></a>pca_prcomp <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(normalized_vsd), <span class="at">center =</span> T, <span class="at">scale. =</span> F) <span class="co">#PCA analysis</span></span>
<span id="cb31-7"><a href="#cb31-7"></a></span>
<span id="cb31-8"><a href="#cb31-8"></a><span class="fu">names</span>(pca_prcomp)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sdev"     "rotation" "center"   "scale"    "x"       </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>pca_prcomp<span class="sc">$</span>x</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         PC1       PC2       PC3           PC4
01_SW_sgCTRL_Norm  -15.07327  2.234704 -1.602628  9.567716e-14
02_SW_sgCTRL_Norm  -14.90781 -2.238765  1.619684 -4.871971e-15
17_SW_sgRELB_3_Hyp  14.92560 -1.909982 -1.893329  1.475209e-14
18_SW_sgRELB_3_Hyp  15.05548  1.914043  1.876273  1.168549e-13</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>PC1_PC2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb35-2"><a href="#cb35-2"></a>  <span class="at">PC1 =</span> pca_prcomp<span class="sc">$</span>x[,<span class="dv">1</span>],</span>
<span id="cb35-3"><a href="#cb35-3"></a>  <span class="at">PC2 =</span> pca_prcomp<span class="sc">$</span>x[,<span class="dv">2</span>],</span>
<span id="cb35-4"><a href="#cb35-4"></a>  <span class="at">type =</span> <span class="fu">rownames</span>(pca_prcomp<span class="sc">$</span>x)</span>
<span id="cb35-5"><a href="#cb35-5"></a>)</span>
<span id="cb35-6"><a href="#cb35-6"></a></span>
<span id="cb35-7"><a href="#cb35-7"></a>var_explained <span class="ot">&lt;-</span> (pca_prcomp<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> <span class="fu">sum</span>(pca_prcomp<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb35-8"><a href="#cb35-8"></a>pc1_percent <span class="ot">&lt;-</span> <span class="fu">round</span>(var_explained[<span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb35-9"><a href="#cb35-9"></a>pc2_percent <span class="ot">&lt;-</span> <span class="fu">round</span>(var_explained[<span class="dv">2</span>] <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb35-10"><a href="#cb35-10"></a></span>
<span id="cb35-11"><a href="#cb35-11"></a><span class="fu">ggplot</span>(PC1_PC2, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">col =</span> type)) <span class="sc">+</span></span>
<span id="cb35-12"><a href="#cb35-12"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb35-13"><a href="#cb35-13"></a>  <span class="fu">geom_text_repel</span>(</span>
<span id="cb35-14"><a href="#cb35-14"></a>    <span class="fu">aes</span>(<span class="at">label =</span> type),</span>
<span id="cb35-15"><a href="#cb35-15"></a>    <span class="at">box.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb35-16"><a href="#cb35-16"></a>    <span class="at">point.padding =</span> <span class="fl">0.3</span>,</span>
<span id="cb35-17"><a href="#cb35-17"></a>    <span class="at">segment.size =</span> <span class="fl">0.3</span>,</span>
<span id="cb35-18"><a href="#cb35-18"></a>    <span class="at">min.segment.length =</span> <span class="dv">0</span></span>
<span id="cb35-19"><a href="#cb35-19"></a>  ) <span class="sc">+</span></span>
<span id="cb35-20"><a href="#cb35-20"></a>  </span>
<span id="cb35-21"><a href="#cb35-21"></a>  <span class="fu">labs</span>(</span>
<span id="cb35-22"><a href="#cb35-22"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, pc1_percent, <span class="st">"% variance)"</span>),</span>
<span id="cb35-23"><a href="#cb35-23"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, pc2_percent, <span class="st">"% variance)"</span>)</span>
<span id="cb35-24"><a href="#cb35-24"></a>  ) <span class="sc">+</span></span>
<span id="cb35-25"><a href="#cb35-25"></a>  </span>
<span id="cb35-26"><a href="#cb35-26"></a>  <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">"off"</span>) <span class="sc">+</span> </span>
<span id="cb35-27"><a href="#cb35-27"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb35-28"><a href="#cb35-28"></a>  <span class="fu">theme</span>(</span>
<span id="cb35-29"><a href="#cb35-29"></a>    <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">5</span>),  </span>
<span id="cb35-30"><a href="#cb35-30"></a>    <span class="at">legend.position =</span> <span class="st">"right"</span></span>
<span id="cb35-31"><a href="#cb35-31"></a>  )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2DE_Heatmap_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It is not exactly the same, what‚Äôs going on?</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>?plotPCA <span class="co">#using the top 500 most variable genes</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="make-a-perfect-heatmap" class="level2">
<h2 class="anchored" data-anchor-id="make-a-perfect-heatmap">Make a perfect heatmap</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>significant_mat <span class="ot">&lt;-</span> normalized_vsd[significant_genes, ]</span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="fu">dim</span>(significant_mat)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 967   4</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(significant_mat))))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2DE_Heatmap_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You get this perfect looking heatmap because you select the genes that are different. So, no surprise at all!</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>metaframe</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 √ó 1
  condition
* &lt;chr&gt;    
1 normoxia 
2 normoxia 
3 hypoxia  
4 hypoxia  </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>meta_annot <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">df =</span> metaframe,</span>
<span id="cb42-2"><a href="#cb42-2"></a>                                <span class="at">col =</span> <span class="fu">list</span>(<span class="at">condition =</span> <span class="fu">c</span>(<span class="st">"hypoxia"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"normoxia"</span> <span class="ot">=</span> <span class="st">"blue"</span>)))</span>
<span id="cb42-3"><a href="#cb42-3"></a></span>
<span id="cb42-4"><a href="#cb42-4"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(significant_mat))),</span>
<span id="cb42-5"><a href="#cb42-5"></a>        <span class="at">top_annotation =</span> meta_annot,</span>
<span id="cb42-6"><a href="#cb42-6"></a>        <span class="at">show_row_names =</span> F,</span>
<span id="cb42-7"><a href="#cb42-7"></a>        <span class="at">name =</span> <span class="st">"scaled normalized</span><span class="sc">\n</span><span class="st">expression"</span></span>
<span id="cb42-8"><a href="#cb42-8"></a>        )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2DE_Heatmap_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>why scaling is important?</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="fu">Heatmap</span>(significant_mat, </span>
<span id="cb43-2"><a href="#cb43-2"></a>        <span class="at">top_annotation =</span> meta_annot,</span>
<span id="cb43-3"><a href="#cb43-3"></a>        <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb43-4"><a href="#cb43-4"></a>        <span class="at">name =</span> <span class="st">"scaled normalized</span><span class="sc">\n</span><span class="st">expression"</span></span>
<span id="cb43-5"><a href="#cb43-5"></a>        )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="2DE_Heatmap_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<!-- theme toggle (top-right) -->
<button id="theme-toggle" style="
  position: fixed;
  top: 15px;
  right: 15px;
  z-index: 9999;
  background: #232b3a;
  color: #d4d4d4;
  border: 1px solid #3d4758;
  padding: 6px 12px;
  border-radius: 6px;
  cursor: pointer;
  font-size: 0.8rem;
">
  üåô Dark Mode
</button>

<!-- sidebar toggle (top-left) -->
<button id="sidebar-toggle" title="Toggle sidebar" style="
  position: fixed;
  top: 15px;
  left: 15px;
  z-index: 9999;
  background: #232b3a;
  color: #d4d4d4;
  border: 1px solid #3d4758;
  padding: 6px 8px;
  border-radius: 6px;
  cursor: pointer;
  font-size: 0.9rem;
">
  ‚ò∞
</button>

<!-- Script con percorso dalla root -->
<script src="../../../toggle.js"></script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/Bioyako\.github\.io\/LearningByDoing\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb44" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb44-1"><a href="#cb44-1"></a><span class="co">---</span></span>
<span id="cb44-2"><a href="#cb44-2"></a><span class="an">title:</span><span class="co"> "2DE_Heatmap"</span></span>
<span id="cb44-3"><a href="#cb44-3"></a><span class="an">author:</span><span class="co"> "Jaakko"</span></span>
<span id="cb44-4"><a href="#cb44-4"></a><span class="an">date:</span><span class="co"> "2025-11-20"</span></span>
<span id="cb44-5"><a href="#cb44-5"></a><span class="an">format:</span></span>
<span id="cb44-6"><a href="#cb44-6"></a><span class="co">  html:</span></span>
<span id="cb44-7"><a href="#cb44-7"></a><span class="co">    css: /styles.css</span></span>
<span id="cb44-8"><a href="#cb44-8"></a><span class="co">    code-copy: true </span></span>
<span id="cb44-9"><a href="#cb44-9"></a><span class="co">    code-line-numbers: true</span></span>
<span id="cb44-10"><a href="#cb44-10"></a><span class="co">    code-overflow: scroll</span></span>
<span id="cb44-11"><a href="#cb44-11"></a><span class="co">    code-tools: true</span></span>
<span id="cb44-12"><a href="#cb44-12"></a><span class="co">    highlight-style: github-dark</span></span>
<span id="cb44-13"><a href="#cb44-13"></a><span class="an">engine:</span><span class="co"> knitr</span></span>
<span id="cb44-14"><a href="#cb44-14"></a><span class="an">execute:</span></span>
<span id="cb44-15"><a href="#cb44-15"></a><span class="co">  echo: true</span></span>
<span id="cb44-16"><a href="#cb44-16"></a><span class="co">  eval: true</span></span>
<span id="cb44-17"><a href="#cb44-17"></a><span class="an">conda:</span></span>
<span id="cb44-18"><a href="#cb44-18"></a><span class="co">  environment: bioinfo #bioinfo_environment.yml</span></span>
<span id="cb44-19"><a href="#cb44-19"></a><span class="co">---</span></span>
<span id="cb44-20"><a href="#cb44-20"></a></span>
<span id="cb44-21"><a href="#cb44-21"></a><span class="fu">## Quarto</span></span>
<span id="cb44-22"><a href="#cb44-22"></a></span>
<span id="cb44-23"><a href="#cb44-23"></a>This is a Quarto document. Quarto is a multi-language, next-generation version of R Markdown. Check the <span class="co">[</span><span class="ot">tutorial</span><span class="co">](https://quarto.org/docs/get-started/hello/rstudio.html)</span>.\</span>
<span id="cb44-24"><a href="#cb44-24"></a>N.B. - Learn the IT foundations to master bioinformatics tools and softwares, and prevent getting tangled on the infamous "dependency hell". Ask for the "Foundations of Bioinformatics Infrastructure".</span>
<span id="cb44-25"><a href="#cb44-25"></a></span>
<span id="cb44-26"><a href="#cb44-26"></a><span class="fu">## Create and activate the environment "bioinfo"</span></span>
<span id="cb44-27"><a href="#cb44-27"></a></span>
<span id="cb44-30"><a href="#cb44-30"></a><span class="in">```{bash}</span></span>
<span id="cb44-31"><a href="#cb44-31"></a><span class="co">#| eval: false</span></span>
<span id="cb44-32"><a href="#cb44-32"></a><span class="ex">mamba</span> create <span class="at">-n</span> bioinfo python=3.11</span>
<span id="cb44-33"><a href="#cb44-33"></a><span class="ex">mamba</span> activate bioinfo</span>
<span id="cb44-34"><a href="#cb44-34"></a><span class="in">```</span></span>
<span id="cb44-35"><a href="#cb44-35"></a></span>
<span id="cb44-36"><a href="#cb44-36"></a><span class="fu">## Download the RNA-seq dataset</span></span>
<span id="cb44-37"><a href="#cb44-37"></a></span>
<span id="cb44-38"><a href="#cb44-38"></a>Let's use a real example - <span class="co">[</span><span class="ot">*RNA-seq of ITPR3 and RELB knockout in SW480 under normoxia and hypoxia*</span><span class="co">]</span>{.underline} (*Homo sapiens*) <span class="ot">&lt;https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197576&gt;</span>.</span>
<span id="cb44-39"><a href="#cb44-39"></a></span>
<span id="cb44-40"><a href="#cb44-40"></a><span class="al">![](images/IM1.png)</span></span>
<span id="cb44-41"><a href="#cb44-41"></a></span>
<span id="cb44-42"><a href="#cb44-42"></a>N.B. - Structure properly your folder ('mkdir folder'):\</span>
<span id="cb44-43"><a href="#cb44-43"></a></span>
<span id="cb44-46"><a href="#cb44-46"></a><span class="in">```{bash}</span></span>
<span id="cb44-47"><a href="#cb44-47"></a><span class="co">#| echo: false</span></span>
<span id="cb44-48"><a href="#cb44-48"></a><span class="bu">cd</span> ..</span>
<span id="cb44-49"><a href="#cb44-49"></a><span class="bu">printf</span> <span class="st">".\n‚îú‚îÄ‚îÄ data/\n‚îú‚îÄ‚îÄ scripts/\n‚îî‚îÄ‚îÄ results/\n"</span></span>
<span id="cb44-50"><a href="#cb44-50"></a><span class="in">```</span></span>
<span id="cb44-51"><a href="#cb44-51"></a></span>
<span id="cb44-52"><a href="#cb44-52"></a>How to download the files from FTP with UNIX terminal: <span class="ot">&lt;https://www.ncbi.nlm.nih.gov/geo/info/download.html&gt;</span>\</span>
<span id="cb44-53"><a href="#cb44-53"></a><span class="ot">&lt;https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/&gt;</span></span>
<span id="cb44-54"><a href="#cb44-54"></a></span>
<span id="cb44-57"><a href="#cb44-57"></a><span class="in">```{bash}</span></span>
<span id="cb44-58"><a href="#cb44-58"></a><span class="co">#| eval: false</span></span>
<span id="cb44-59"><a href="#cb44-59"></a><span class="co"># Move to /data folder !!!</span></span>
<span id="cb44-60"><a href="#cb44-60"></a><span class="bu">cd</span> data</span>
<span id="cb44-61"><a href="#cb44-61"></a><span class="fu">wget</span> https://ftp.ncbi.nlm.nih.gov/geo/series/GSE197nnn/GSE197576/suppl/GSE197576_raw_gene_counts_matrix.tsv.gz</span>
<span id="cb44-62"><a href="#cb44-62"></a><span class="in">```</span></span>
<span id="cb44-63"><a href="#cb44-63"></a></span>
<span id="cb44-64"><a href="#cb44-64"></a>Alternative use GEOquery <span class="ot">&lt;https://bioconductor.org/packages/release/bioc/html/GEOquery.html&gt;</span>.</span>
<span id="cb44-65"><a href="#cb44-65"></a></span>
<span id="cb44-66"><a href="#cb44-66"></a><span class="fu">## Check the data ad extract the samples of interest</span></span>
<span id="cb44-67"><a href="#cb44-67"></a></span>
<span id="cb44-68"><a href="#cb44-68"></a>Check the data in the compressed RNA-seq file.</span>
<span id="cb44-69"><a href="#cb44-69"></a></span>
<span id="cb44-72"><a href="#cb44-72"></a><span class="in">```{bash}</span></span>
<span id="cb44-73"><a href="#cb44-73"></a><span class="co">#| eval: false</span></span>
<span id="cb44-74"><a href="#cb44-74"></a><span class="fu">zless</span> <span class="at">-S</span> GSE197576_raw_gene_counts_matrix.tsv.gz</span>
<span id="cb44-75"><a href="#cb44-75"></a><span class="co"># z because is compressed</span></span>
<span id="cb44-76"><a href="#cb44-76"></a><span class="co"># less is the faster command to view the conte of a file</span></span>
<span id="cb44-77"><a href="#cb44-77"></a><span class="co"># -S maintain one raw column</span></span>
<span id="cb44-78"><a href="#cb44-78"></a><span class="co"># press q to quit</span></span>
<span id="cb44-79"><a href="#cb44-79"></a><span class="in">```</span></span>
<span id="cb44-80"><a href="#cb44-80"></a></span>
<span id="cb44-81"><a href="#cb44-81"></a>Install csvtk annd check the data in a pretty aligned manner.</span>
<span id="cb44-82"><a href="#cb44-82"></a></span>
<span id="cb44-85"><a href="#cb44-85"></a><span class="in">```{bash}</span></span>
<span id="cb44-86"><a href="#cb44-86"></a><span class="co">#| eval: false</span></span>
<span id="cb44-87"><a href="#cb44-87"></a><span class="ex">mamba</span> install <span class="at">-c</span> bioconda csvtk</span>
<span id="cb44-88"><a href="#cb44-88"></a><span class="in">```</span></span>
<span id="cb44-89"><a href="#cb44-89"></a></span>
<span id="cb44-92"><a href="#cb44-92"></a><span class="in">```{bash}</span></span>
<span id="cb44-93"><a href="#cb44-93"></a><span class="co">#| eval: false</span></span>
<span id="cb44-94"><a href="#cb44-94"></a><span class="ex">csvtk</span> pretty <span class="at">-t</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span>
<span id="cb44-95"><a href="#cb44-95"></a><span class="co"># -t tab the limit</span></span>
<span id="cb44-96"><a href="#cb44-96"></a><span class="co"># press q to quit</span></span>
<span id="cb44-97"><a href="#cb44-97"></a><span class="in">```</span></span>
<span id="cb44-98"><a href="#cb44-98"></a></span>
<span id="cb44-101"><a href="#cb44-101"></a><span class="in">```{bash}</span></span>
<span id="cb44-102"><a href="#cb44-102"></a><span class="co">#| echo: true</span></span>
<span id="cb44-103"><a href="#cb44-103"></a><span class="ex">csvtk</span> headers <span class="at">-t</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz </span>
<span id="cb44-104"><a href="#cb44-104"></a><span class="in">```</span></span>
<span id="cb44-105"><a href="#cb44-105"></a></span>
<span id="cb44-106"><a href="#cb44-106"></a>N.B - Get csvtk at <span class="ot">&lt;https://github.com/shenwei356/csvtk&gt;</span></span>
<span id="cb44-107"><a href="#cb44-107"></a></span>
<span id="cb44-108"><a href="#cb44-108"></a>Extract the gene expression values from:</span>
<span id="cb44-109"><a href="#cb44-109"></a></span>
<span id="cb44-110"><a href="#cb44-110"></a><span class="pp">|</span>  <span class="pp">|</span>  <span class="pp">|</span>  <span class="pp">|</span>  <span class="pp">|</span></span>
<span id="cb44-111"><a href="#cb44-111"></a><span class="pp">|------------------|------------------|------------------|------------------|</span></span>
<span id="cb44-112"><a href="#cb44-112"></a><span class="pp">|</span> *01_SW_sgCTRL_Norm* | *02_SW_sgCTRL_Norm* | *17_SW_sgRELB_3_Hyp* | *18_SW_sgRELB_3_Hyp* <span class="pp">|</span></span>
<span id="cb44-113"><a href="#cb44-113"></a></span>
<span id="cb44-116"><a href="#cb44-116"></a><span class="in">```{bash}</span></span>
<span id="cb44-117"><a href="#cb44-117"></a><span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-f1,2,3,12,13</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="kw">|</span> <span class="fu">head</span></span>
<span id="cb44-118"><a href="#cb44-118"></a><span class="in">```</span></span>
<span id="cb44-119"><a href="#cb44-119"></a></span>
<span id="cb44-122"><a href="#cb44-122"></a><span class="in">```{bash}</span></span>
<span id="cb44-123"><a href="#cb44-123"></a><span class="co">#| eval: false</span></span>
<span id="cb44-124"><a href="#cb44-124"></a><span class="ex">csvtk</span> cut <span class="at">-t</span> <span class="at">-f1,1,2,12,13</span> ../data/GSE197576_raw_gene_counts_matrix.tsv.gz <span class="op">&gt;</span> raw.counts.tsv</span>
<span id="cb44-125"><a href="#cb44-125"></a><span class="in">```</span></span>
<span id="cb44-126"><a href="#cb44-126"></a></span>
<span id="cb44-127"><a href="#cb44-127"></a>Your */data* folder now contains:</span>
<span id="cb44-128"><a href="#cb44-128"></a></span>
<span id="cb44-131"><a href="#cb44-131"></a><span class="in">```{bash}</span></span>
<span id="cb44-132"><a href="#cb44-132"></a><span class="co">#| echo: false</span></span>
<span id="cb44-133"><a href="#cb44-133"></a><span class="bu">cd</span> ..</span>
<span id="cb44-134"><a href="#cb44-134"></a><span class="ex">tree</span> <span class="at">-n</span> <span class="at">-F</span> <span class="at">-L</span> 1 <span class="at">--noreport</span> data </span>
<span id="cb44-135"><a href="#cb44-135"></a><span class="in">```</span></span>
<span id="cb44-136"><a href="#cb44-136"></a></span>
<span id="cb44-137"><a href="#cb44-137"></a><span class="fu">## Read the data into R and make a DESeq2 object</span></span>
<span id="cb44-138"><a href="#cb44-138"></a></span>
<span id="cb44-139"><a href="#cb44-139"></a>Follow the tutorial <span class="ot">&lt;http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html&gt;</span>.</span>
<span id="cb44-140"><a href="#cb44-140"></a></span>
<span id="cb44-141"><a href="#cb44-141"></a><span class="fu">### Setup</span></span>
<span id="cb44-142"><a href="#cb44-142"></a></span>
<span id="cb44-143"><a href="#cb44-143"></a>Install the R-packages for this project within the 'bioinfo' virtual environment.</span>
<span id="cb44-144"><a href="#cb44-144"></a></span>
<span id="cb44-147"><a href="#cb44-147"></a><span class="in">```{bash}</span></span>
<span id="cb44-148"><a href="#cb44-148"></a><span class="co">#| eval: false</span></span>
<span id="cb44-149"><a href="#cb44-149"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-tidyverse</span>
<span id="cb44-150"><a href="#cb44-150"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-dplyr</span>
<span id="cb44-151"><a href="#cb44-151"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-here</span>
<span id="cb44-152"><a href="#cb44-152"></a><span class="ex">mamba</span> install <span class="at">-c</span> bioconda bioconductor-deseq2</span>
<span id="cb44-153"><a href="#cb44-153"></a><span class="ex">mamba</span> intall <span class="at">-c</span> conda-forge r-ggplot2</span>
<span id="cb44-154"><a href="#cb44-154"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-ggrepel</span>
<span id="cb44-155"><a href="#cb44-155"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-BiocManager   </span>
<span id="cb44-156"><a href="#cb44-156"></a><span class="in">```</span></span>
<span id="cb44-157"><a href="#cb44-157"></a></span>
<span id="cb44-158"><a href="#cb44-158"></a>Save the environment for reproducibility.</span>
<span id="cb44-159"><a href="#cb44-159"></a></span>
<span id="cb44-162"><a href="#cb44-162"></a><span class="in">```{bash}</span></span>
<span id="cb44-163"><a href="#cb44-163"></a><span class="co">#| eval: false</span></span>
<span id="cb44-164"><a href="#cb44-164"></a><span class="ex">mamba</span> env export <span class="at">--no-builds</span> <span class="op">&gt;</span> bioinfo_environment.yml</span>
<span id="cb44-165"><a href="#cb44-165"></a><span class="in">```</span></span>
<span id="cb44-166"><a href="#cb44-166"></a></span>
<span id="cb44-167"><a href="#cb44-167"></a>Open RStudio from the terminal: you are now working with the R software within 'bioinfo' and you can find the packages previously installed.</span>
<span id="cb44-168"><a href="#cb44-168"></a></span>
<span id="cb44-171"><a href="#cb44-171"></a><span class="in">```{bash}</span></span>
<span id="cb44-172"><a href="#cb44-172"></a><span class="co">#| eval: false</span></span>
<span id="cb44-173"><a href="#cb44-173"></a><span class="ex">open</span> <span class="at">-na</span> RStudio</span>
<span id="cb44-174"><a href="#cb44-174"></a><span class="in">```</span></span>
<span id="cb44-175"><a href="#cb44-175"></a></span>
<span id="cb44-176"><a href="#cb44-176"></a><span class="fu">### Let‚Äôs start</span></span>
<span id="cb44-177"><a href="#cb44-177"></a></span>
<span id="cb44-180"><a href="#cb44-180"></a><span class="in">```{r}</span></span>
<span id="cb44-181"><a href="#cb44-181"></a><span class="co">#| message: false</span></span>
<span id="cb44-182"><a href="#cb44-182"></a><span class="co">#| warning: false</span></span>
<span id="cb44-183"><a href="#cb44-183"></a><span class="fu">library</span>(dplyr) <span class="co">#https://dplyr.tidyverse.org/</span></span>
<span id="cb44-184"><a href="#cb44-184"></a><span class="fu">library</span>(readr) <span class="co">#https://readr.tidyverse.org/</span></span>
<span id="cb44-185"><a href="#cb44-185"></a><span class="fu">library</span>(here) <span class="co">#https://cran.r-project.org/web/packages/here/vignettes/here.html</span></span>
<span id="cb44-186"><a href="#cb44-186"></a><span class="fu">library</span>(DESeq2) <span class="co">#https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8</span></span>
<span id="cb44-187"><a href="#cb44-187"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb44-188"><a href="#cb44-188"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb44-189"><a href="#cb44-189"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"ComplexHeatmap"</span>)</span>
<span id="cb44-190"><a href="#cb44-190"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb44-191"><a href="#cb44-191"></a></span>
<span id="cb44-192"><a href="#cb44-192"></a>raw_counts <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="fu">here</span>(<span class="st">"01_Heatmap_Genomics/2_DE_Heatmap/data/raw.counts.tsv"</span>)) <span class="co">#import the dataset</span></span>
<span id="cb44-193"><a href="#cb44-193"></a></span>
<span id="cb44-194"><a href="#cb44-194"></a>raw_counts_mat <span class="ot">&lt;-</span> raw_counts[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">%&gt;%</span> as.matrix <span class="co">#create the values matrix removing the first column for DESeq2</span></span>
<span id="cb44-195"><a href="#cb44-195"></a><span class="fu">head</span>(raw_counts_mat)</span>
<span id="cb44-196"><a href="#cb44-196"></a></span>
<span id="cb44-197"><a href="#cb44-197"></a><span class="fu">rownames</span>(raw_counts_mat) <span class="ot">&lt;-</span> raw_counts<span class="sc">$</span>gene <span class="co">#add row names to the matrix</span></span>
<span id="cb44-198"><a href="#cb44-198"></a><span class="fu">head</span>(raw_counts_mat)</span>
<span id="cb44-199"><a href="#cb44-199"></a><span class="in">```</span></span>
<span id="cb44-200"><a href="#cb44-200"></a></span>
<span id="cb44-201"><a href="#cb44-201"></a>Make a dataframe for the metadata.</span>
<span id="cb44-202"><a href="#cb44-202"></a></span>
<span id="cb44-205"><a href="#cb44-205"></a><span class="in">```{r}</span></span>
<span id="cb44-206"><a href="#cb44-206"></a><span class="co">#| message: false</span></span>
<span id="cb44-207"><a href="#cb44-207"></a><span class="co">#| warning: false</span></span>
<span id="cb44-208"><a href="#cb44-208"></a>metaframe <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">condition =</span> <span class="fu">c</span> (<span class="st">"normoxia"</span>, <span class="st">"normoxia"</span>, <span class="st">"hypoxia"</span>, <span class="st">"hypoxia"</span>))</span>
<span id="cb44-209"><a href="#cb44-209"></a><span class="fu">rownames</span>(metaframe) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(raw_counts_mat)</span>
<span id="cb44-210"><a href="#cb44-210"></a>metaframe</span>
<span id="cb44-211"><a href="#cb44-211"></a><span class="in">```</span></span>
<span id="cb44-212"><a href="#cb44-212"></a></span>
<span id="cb44-213"><a href="#cb44-213"></a>Make a DESeq2 object (<span class="co">[</span><span class="ot">tutorial</span><span class="co">](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)</span>).</span>
<span id="cb44-214"><a href="#cb44-214"></a></span>
<span id="cb44-217"><a href="#cb44-217"></a><span class="in">```{r}</span></span>
<span id="cb44-218"><a href="#cb44-218"></a><span class="co">#| message: false</span></span>
<span id="cb44-219"><a href="#cb44-219"></a><span class="co">#| warning: false</span></span>
<span id="cb44-220"><a href="#cb44-220"></a><span class="fu">all</span>(<span class="fu">rownames</span>(metaframe) <span class="sc">==</span> <span class="fu">colnames</span>(raw_counts_mat)) <span class="co">#check</span></span>
<span id="cb44-221"><a href="#cb44-221"></a></span>
<span id="cb44-222"><a href="#cb44-222"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> raw_counts_mat, <span class="co">#input object with design</span></span>
<span id="cb44-223"><a href="#cb44-223"></a>                              <span class="at">colData =</span> metaframe,</span>
<span id="cb44-224"><a href="#cb44-224"></a>                              <span class="at">design =</span> <span class="sc">~</span> condition)</span>
<span id="cb44-225"><a href="#cb44-225"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds) <span class="co">#DE analysis</span></span>
<span id="cb44-226"><a href="#cb44-226"></a></span>
<span id="cb44-227"><a href="#cb44-227"></a>?results</span>
<span id="cb44-228"><a href="#cb44-228"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"condition"</span>, <span class="st">"hypoxia"</span>, <span class="st">"normoxia"</span>)) <span class="co">#extract results from the analysis </span></span>
<span id="cb44-229"><a href="#cb44-229"></a></span>
<span id="cb44-230"><a href="#cb44-230"></a>res <span class="sc">%&gt;%</span> </span>
<span id="cb44-231"><a href="#cb44-231"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb44-232"><a href="#cb44-232"></a>  <span class="fu">arrange</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.1</span>, <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;=</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="co">#filter significant and substantial expression changes </span></span>
<span id="cb44-233"><a href="#cb44-233"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb44-234"><a href="#cb44-234"></a></span>
<span id="cb44-235"><a href="#cb44-235"></a>significant_genes <span class="ot">&lt;-</span> res <span class="sc">%&gt;%</span> <span class="co">#identify the respective genes</span></span>
<span id="cb44-236"><a href="#cb44-236"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb44-237"><a href="#cb44-237"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;=</span> <span class="fl">0.1</span>, <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;=</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-238"><a href="#cb44-238"></a>  <span class="fu">rownames</span>()</span>
<span id="cb44-239"><a href="#cb44-239"></a></span>
<span id="cb44-240"><a href="#cb44-240"></a>significant_genes[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb44-241"><a href="#cb44-241"></a><span class="in">```</span></span>
<span id="cb44-242"><a href="#cb44-242"></a></span>
<span id="cb44-243"><a href="#cb44-243"></a><span class="fu">### PCA Analysis to verify your results</span></span>
<span id="cb44-244"><a href="#cb44-244"></a></span>
<span id="cb44-247"><a href="#cb44-247"></a><span class="in">```{r}</span></span>
<span id="cb44-248"><a href="#cb44-248"></a><span class="co">#| message: true</span></span>
<span id="cb44-249"><a href="#cb44-249"></a><span class="co">#| warning: false</span></span>
<span id="cb44-250"><a href="#cb44-250"></a>vsd <span class="ot">&lt;-</span> <span class="fu">vst</span>(dds, <span class="at">blind =</span> F) <span class="co">#normalization applying a variance stabilizing transformation (VST)</span></span>
<span id="cb44-251"><a href="#cb44-251"></a></span>
<span id="cb44-252"><a href="#cb44-252"></a>p <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(vsd, <span class="at">intgroup =</span> <span class="fu">c</span>(<span class="st">"condition"</span>))</span>
<span id="cb44-253"><a href="#cb44-253"></a>p <span class="sc">+</span> <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.7</span>, <span class="fl">0.7</span>))</span>
<span id="cb44-254"><a href="#cb44-254"></a><span class="in">```</span></span>
<span id="cb44-255"><a href="#cb44-255"></a></span>
<span id="cb44-256"><a href="#cb44-256"></a>You can make a PCA by yourself.</span>
<span id="cb44-257"><a href="#cb44-257"></a></span>
<span id="cb44-260"><a href="#cb44-260"></a><span class="in">```{r}</span></span>
<span id="cb44-261"><a href="#cb44-261"></a><span class="co">#| warning: false</span></span>
<span id="cb44-262"><a href="#cb44-262"></a><span class="co"># vsd &lt;- vst(dds, blind = F)</span></span>
<span id="cb44-263"><a href="#cb44-263"></a><span class="fu">head</span>(<span class="fu">assay</span>(vsd), <span class="dv">3</span>)</span>
<span id="cb44-264"><a href="#cb44-264"></a>?assay</span>
<span id="cb44-265"><a href="#cb44-265"></a></span>
<span id="cb44-266"><a href="#cb44-266"></a>normalized_vsd <span class="ot">&lt;-</span> <span class="fu">assay</span>(vsd) <span class="sc">%&gt;%</span> <span class="co">#from DeSeq object to matrix</span></span>
<span id="cb44-267"><a href="#cb44-267"></a> <span class="fu">as.matrix</span>()</span>
<span id="cb44-268"><a href="#cb44-268"></a></span>
<span id="cb44-269"><a href="#cb44-269"></a>pca_prcomp <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(normalized_vsd), <span class="at">center =</span> T, <span class="at">scale. =</span> F) <span class="co">#PCA analysis</span></span>
<span id="cb44-270"><a href="#cb44-270"></a></span>
<span id="cb44-271"><a href="#cb44-271"></a><span class="fu">names</span>(pca_prcomp)</span>
<span id="cb44-272"><a href="#cb44-272"></a>pca_prcomp<span class="sc">$</span>x</span>
<span id="cb44-273"><a href="#cb44-273"></a></span>
<span id="cb44-274"><a href="#cb44-274"></a>PC1_PC2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb44-275"><a href="#cb44-275"></a>  <span class="at">PC1 =</span> pca_prcomp<span class="sc">$</span>x[,<span class="dv">1</span>],</span>
<span id="cb44-276"><a href="#cb44-276"></a>  <span class="at">PC2 =</span> pca_prcomp<span class="sc">$</span>x[,<span class="dv">2</span>],</span>
<span id="cb44-277"><a href="#cb44-277"></a>  <span class="at">type =</span> <span class="fu">rownames</span>(pca_prcomp<span class="sc">$</span>x)</span>
<span id="cb44-278"><a href="#cb44-278"></a>)</span>
<span id="cb44-279"><a href="#cb44-279"></a></span>
<span id="cb44-280"><a href="#cb44-280"></a>var_explained <span class="ot">&lt;-</span> (pca_prcomp<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> <span class="fu">sum</span>(pca_prcomp<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb44-281"><a href="#cb44-281"></a>pc1_percent <span class="ot">&lt;-</span> <span class="fu">round</span>(var_explained[<span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb44-282"><a href="#cb44-282"></a>pc2_percent <span class="ot">&lt;-</span> <span class="fu">round</span>(var_explained[<span class="dv">2</span>] <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb44-283"><a href="#cb44-283"></a></span>
<span id="cb44-284"><a href="#cb44-284"></a><span class="fu">ggplot</span>(PC1_PC2, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">col =</span> type)) <span class="sc">+</span></span>
<span id="cb44-285"><a href="#cb44-285"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb44-286"><a href="#cb44-286"></a>  <span class="fu">geom_text_repel</span>(</span>
<span id="cb44-287"><a href="#cb44-287"></a>    <span class="fu">aes</span>(<span class="at">label =</span> type),</span>
<span id="cb44-288"><a href="#cb44-288"></a>    <span class="at">box.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb44-289"><a href="#cb44-289"></a>    <span class="at">point.padding =</span> <span class="fl">0.3</span>,</span>
<span id="cb44-290"><a href="#cb44-290"></a>    <span class="at">segment.size =</span> <span class="fl">0.3</span>,</span>
<span id="cb44-291"><a href="#cb44-291"></a>    <span class="at">min.segment.length =</span> <span class="dv">0</span></span>
<span id="cb44-292"><a href="#cb44-292"></a>  ) <span class="sc">+</span></span>
<span id="cb44-293"><a href="#cb44-293"></a>  </span>
<span id="cb44-294"><a href="#cb44-294"></a>  <span class="fu">labs</span>(</span>
<span id="cb44-295"><a href="#cb44-295"></a>    <span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, pc1_percent, <span class="st">"% variance)"</span>),</span>
<span id="cb44-296"><a href="#cb44-296"></a>    <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, pc2_percent, <span class="st">"% variance)"</span>)</span>
<span id="cb44-297"><a href="#cb44-297"></a>  ) <span class="sc">+</span></span>
<span id="cb44-298"><a href="#cb44-298"></a>  </span>
<span id="cb44-299"><a href="#cb44-299"></a>  <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">"off"</span>) <span class="sc">+</span> </span>
<span id="cb44-300"><a href="#cb44-300"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb44-301"><a href="#cb44-301"></a>  <span class="fu">theme</span>(</span>
<span id="cb44-302"><a href="#cb44-302"></a>    <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">5</span>),  </span>
<span id="cb44-303"><a href="#cb44-303"></a>    <span class="at">legend.position =</span> <span class="st">"right"</span></span>
<span id="cb44-304"><a href="#cb44-304"></a>  )</span>
<span id="cb44-305"><a href="#cb44-305"></a><span class="in">```</span></span>
<span id="cb44-306"><a href="#cb44-306"></a></span>
<span id="cb44-307"><a href="#cb44-307"></a>It is not exactly the same, what's going on?</span>
<span id="cb44-308"><a href="#cb44-308"></a></span>
<span id="cb44-311"><a href="#cb44-311"></a><span class="in">```{r}</span></span>
<span id="cb44-312"><a href="#cb44-312"></a><span class="co">#| eval: false</span></span>
<span id="cb44-313"><a href="#cb44-313"></a>?plotPCA <span class="co">#using the top 500 most variable genes</span></span>
<span id="cb44-314"><a href="#cb44-314"></a><span class="in">```</span></span>
<span id="cb44-315"><a href="#cb44-315"></a></span>
<span id="cb44-316"><a href="#cb44-316"></a><span class="fu">## Make a perfect heatmap</span></span>
<span id="cb44-317"><a href="#cb44-317"></a></span>
<span id="cb44-320"><a href="#cb44-320"></a><span class="in">```{r}</span></span>
<span id="cb44-321"><a href="#cb44-321"></a>significant_mat <span class="ot">&lt;-</span> normalized_vsd[significant_genes, ]</span>
<span id="cb44-322"><a href="#cb44-322"></a><span class="fu">dim</span>(significant_mat)</span>
<span id="cb44-323"><a href="#cb44-323"></a></span>
<span id="cb44-324"><a href="#cb44-324"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(significant_mat))))</span>
<span id="cb44-325"><a href="#cb44-325"></a></span>
<span id="cb44-326"><a href="#cb44-326"></a><span class="in">```</span></span>
<span id="cb44-327"><a href="#cb44-327"></a></span>
<span id="cb44-328"><a href="#cb44-328"></a>You get this perfect looking heatmap because you select the genes that are different. So, no surprise at all!</span>
<span id="cb44-329"><a href="#cb44-329"></a></span>
<span id="cb44-332"><a href="#cb44-332"></a><span class="in">```{r}</span></span>
<span id="cb44-333"><a href="#cb44-333"></a>metaframe</span>
<span id="cb44-334"><a href="#cb44-334"></a></span>
<span id="cb44-335"><a href="#cb44-335"></a>meta_annot <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">df =</span> metaframe,</span>
<span id="cb44-336"><a href="#cb44-336"></a>                                <span class="at">col =</span> <span class="fu">list</span>(<span class="at">condition =</span> <span class="fu">c</span>(<span class="st">"hypoxia"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"normoxia"</span> <span class="ot">=</span> <span class="st">"blue"</span>)))</span>
<span id="cb44-337"><a href="#cb44-337"></a></span>
<span id="cb44-338"><a href="#cb44-338"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(significant_mat))),</span>
<span id="cb44-339"><a href="#cb44-339"></a>        <span class="at">top_annotation =</span> meta_annot,</span>
<span id="cb44-340"><a href="#cb44-340"></a>        <span class="at">show_row_names =</span> F,</span>
<span id="cb44-341"><a href="#cb44-341"></a>        <span class="at">name =</span> <span class="st">"scaled normalized</span><span class="sc">\n</span><span class="st">expression"</span></span>
<span id="cb44-342"><a href="#cb44-342"></a>        )</span>
<span id="cb44-343"><a href="#cb44-343"></a><span class="in">```</span></span>
<span id="cb44-344"><a href="#cb44-344"></a></span>
<span id="cb44-345"><a href="#cb44-345"></a>why scaling is important?</span>
<span id="cb44-346"><a href="#cb44-346"></a></span>
<span id="cb44-349"><a href="#cb44-349"></a><span class="in">```{r}</span></span>
<span id="cb44-350"><a href="#cb44-350"></a><span class="co">#| message: false</span></span>
<span id="cb44-351"><a href="#cb44-351"></a><span class="co">#| warning: false</span></span>
<span id="cb44-352"><a href="#cb44-352"></a><span class="fu">Heatmap</span>(significant_mat, </span>
<span id="cb44-353"><a href="#cb44-353"></a>        <span class="at">top_annotation =</span> meta_annot,</span>
<span id="cb44-354"><a href="#cb44-354"></a>        <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb44-355"><a href="#cb44-355"></a>        <span class="at">name =</span> <span class="st">"scaled normalized</span><span class="sc">\n</span><span class="st">expression"</span></span>
<span id="cb44-356"><a href="#cb44-356"></a>        )</span>
<span id="cb44-357"><a href="#cb44-357"></a><span class="in">```</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>